### PR TITLE
8342206: Convenience method to check if a ClassEntry matches a ClassDesc

### PR DESCRIPTION
**Pull request by liach**
Currently, to efficiently check if a `ClassEntry`, such as one from an `InvokeInstruction`, is of a particular class, we use such a pattern (as seen in JEP 486 [Appendix](https://openjdk.org/jeps/486#Appendix)):
```java
inst.owner().name().equalsString("java/lang/System")
```
This pattern has a few issues:

1. This is not straightforward. Users may be tempted to write
```java
inst.owner().asSymbol().equals(CD_System)
```
unaware of the degraded performance from extra conversions.

2. We aim to reduce the use of "internal names" with encapsulation. Direct use of `"java/lang/System"` goes against this, and we have no constants offered to reduce such error-prone use of internal name.

Thus, I propose a new API in `ClassEntry`:
```java
boolean equalsSymbol(ClassDesc)
```
that can be used conveniently to check the identity of a class entry with a clean syntax and better performance.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires CSR request [JDK-8342210](https://bugs.openjdk.org/browse/JDK-8342210) to be approved

### Issues
 * [JDK-8342206](https://bugs.openjdk.org/browse/JDK-8342206): Convenience method to check if a ClassEntry matches a ClassDesc (**Enhancement** - P4)
 * [JDK-8342210](https://bugs.openjdk.org/browse/JDK-8342210): Convenience method to check if a ClassEntry matches a ClassDesc (**CSR**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21526/head:pull/21526` \
`$ git checkout pull/21526`

Update a local copy of the PR: \
`$ git checkout pull/21526` \
`$ git pull https://git.openjdk.org/jdk.git pull/21526/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21526`

View PR using the GUI difftool: \
`$ git pr show -t 21526`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21526.diff">https://git.openjdk.org/jdk/pull/21526.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21526#issuecomment-2414634629)
</details>

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java [hunk 1]
```diff
@@ -599,6 +599,32 @@ public ClassDesc asSymbol() {
...
         }
 
+        @Override
+        public boolean equalsSymbol(ClassDesc symbol) {
+            if (symbol.isPrimitive()) // implicit null check
+                return false;
+
+            var mySym = this.sym;
+            if (mySym != null)
+                return mySym.equals(symbol);
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:609]: Note that `ClassEntry::asSymbol()` also caches array `ClassDesc`s in `ref1.typeSym` so that the symbol cache can be shared with `FieldRefEntry` and `ConstantDynamicEntry`.

--------------------------------------------------------------------------------

I suggest adding the following check:
```java
if (ref1.typeSym instanceof ClassDesc ref1Sym && ref1Sym.isArray()) {
	return ref1Sym.equals(symbol);
}
```

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:609]: I think adding this on the sym present hot path might be too costly. A better way is to pull in the sym before `ref1.equalsString` and push the compatible symbol back to utf8 if we have `equals`.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AbstractPoolEntry.java:621]: Done.

**Comment by ExE-Boss** [comment]
I think it might be a good idea to also have `equalsSymbol(…)` methods for the other pool entries.

**Comment by liach** [comment]
> I think it might be a good idea to also have `equalsSymbol(…)` methods for the other pool entries.

Indeed, this can avoid promotion to String state for all applicable symbols. I decide to start with the ClassEntry as this is the most error-prone and worth optimization.

**Pull request closed without merging**